Makefile

```diff
@@ -31,40 +32,59 @@ all: docker
 gobuild:
 	./scripts/build false
 
+# create output directories
+.out-stamp:
+	mkdir -p ./out/test-artifacts ./out/cni-plugins
+	touch .out-stamp
+
+out: .out-stamp
+
 # Basic go build
 static:
 	./scripts/build
 
+BUILDER_IMAGE="amazon/amazon-ecs-agent-build:make"
+.builder-image-stamp: scripts/dockerfiles/Dockerfile.build
+	@docker build -f scripts/dockerfiles/Dockerfile.build -t $(BUILDER_IMAGE) .
+	touch .builder-image-stamp
+
+builder-image: .builder-image-stamp
+
 # 'build-in-docker' builds the agent within a dockerfile and saves it to the ./out
 # directory
-build-in-docker:
-	@docker build -f scripts/dockerfiles/Dockerfile.build -t "amazon/amazon-ecs-agent-build:make" .
+# TODO: make this idempotent
+build-in-docker: builder-image out
 	@docker run --net=none \
-	  -e TARGET_OS="${TARGET_OS}" \
-	  -e LDFLAGS="-X github.com/aws/amazon-ecs-agent/agent/config.DefaultPauseContainerTag=$(PAUSE_CONTAINER_TAG) \
-	  -X github.com/aws/amazon-ecs-agent/agent/config.DefaultPauseContainerImageName=$(PAUSE_CONTAINER_IMAGE)" \
-	  -v "$(PWD)/out:/out" \
-	  -v "$(PWD):/go/src/github.com/aws/amazon-ecs-agent" \
-	  "amazon/amazon-ecs-agent-build:make"
+		--env TARGET_OS="${TARGET_OS}" \
+		--env LDFLAGS="-X github.com/aws/amazon-ecs-agent/agent/config.DefaultPauseContainerTag=$(PAUSE_CONTAINER_TAG) \
+			-X github.com/aws/amazon-ecs-agent/agent/config.DefaultPauseContainerImageName=$(PAUSE_CONTAINER_IMAGE)" \
+		--volume "$(PWD)/out:/out" \
+		--volume "$(PWD):/go/src/github.com/aws/amazon-ecs-agent" \
+		--user "$(USERID)" \
```

Comment:
Can you please document why the `--user` param is required here?

Reply:
I'll figure out the best place to put it -- we use it in several places.

The reason we need it is so that the build artifacts are owned by the user running `make` instead of root. This means `make clean` will work.